bluecellulab/graph.py

```diff
@@ -0,0 +1,85 @@
+"""Graph representation of Cells and Synapses."""
+
+from matplotlib.cm import ScalarMappable
+from matplotlib.colors import Normalize
+import matplotlib.pyplot as plt
+import networkx as nx
+import numpy as np
+
+
+from bluecellulab.cell.cell_dict import CellDict
+
+
+def build_graph(cells: CellDict) -> nx.DiGraph:
+    G = nx.DiGraph()
+
+    # Add nodes (cells) to the graph
+    for cell_id, cell in cells.items():
+        G.add_node(cell_id, label=str(cell_id.id), population=cell_id.population_name)
+
+    # Extract and add edges (connections) to the graph from each cell
+    for cell_id, cell in cells.items():
+        for connection in cell.connections.values():
+            # Check if pre_cell exists for the connection
+            if connection.pre_cell is None:
+                continue
+
+            # Source is the pre_cell from the connection
+            source_cell_id = connection.pre_cell.cell_id
+
+            # Target is the post-synapse cell from the connection
+            target_cell_id = connection.post_synapse.post_cell_id
+
+            # Check if both source and target cells are within the current cell collection
+            if source_cell_id in cells and target_cell_id in cells:
+                G.add_edge(source_cell_id, target_cell_id, weight=connection.weight)
+
+    return G
+
+
+def plot_graph(G: nx.Graph, node_size: float = 400, edge_width: float = 0.4, node_distance: float = 1.6):
+    # Extract unique populations from the graph nodes
+    populations = list(set([cell_id.population_name for cell_id in G.nodes()]))
+
+    # Create a color map for each population
+    color_map = plt.cm.tab20(np.linspace(0, 1, len(populations)))  # type: ignore[attr-defined]
+    population_color = dict(zip(populations, color_map))
+
+    # Create node colors based on their population
+    node_colors = [population_color[node.population_name] for node in G.nodes()]
```

Comment:
Here, are we getting the population name property of each node, or are we getting the population of the cell_id in each node? I am asking because at line 42, we are doing the same thing, but the variable is named cell_id.
It would be good to be precise about what we are doing, for readability

Reply:
Actually line 42 is constructing a `set` and removing the duplicates. This line 49 constructs a list with duplicates.

Reply:
Yes, I know. But what I meant, is that even though in line 42 it is named `cell_id`, and here it is named `node`, it looks like it is the same variable we access. In `G`, we have something like this for each node `(cell_id, id, population_name)`, and my question is, when we are accessing `population_name`, are we getting the population_name in the node (the 3rd element in my set), or are we accessing `cell_id.population_name` (via the 1st element in my set).

Reply:
Oh thanks, a good catch. The "label" and "population" attributes of the node are never used in G. There is duplicated information. Instead of this:
`G.add_node(cell_id, label=str(cell_id.id), population=cell_id.population_name)`
we can just write this:
`G.add_node(cell_id)`

Reply:
Always the 1st element in your set is used, the 3rd is never used.

Reply:
Ok, thanks for clarifying

Reply:
Thanks for noticing, let me write a patch